src/lem/constrainer.rs

```diff
@@ -81,145 +116,390 @@ impl LEM {
+                    &mut cs.namespace(|| format!("slot_hash{}_{}", $preimg.len() / 2, $hash_index)),
+                    $preimg.to_vec(),
+                    $constants,
+                )?;
+
+                implies_equal(
+                    &mut cs.namespace(|| {
+                        format!(
+                            "implies equal hash for hash{}_{}",
+                            $preimg.len() / 2,
+                            $hash_index
+                        )
+                    }),
+                    $concrete_path,
+                    $img,
+                    &alloc_hash,
+                )?;
+            };
+        }
+
+        for SlotData {
+            arity,
+            concrete_path,
+            preimg,
+            img,
+        } in slots_data
+        {
+            // First we constrain concrete path hashes, later we
+            // fill with dummies in order to always use the same number of hashes.
+            if Self::on_concrete_path(concrete_path)? {
+                match arity {
+                    HashArity::A2 => {
+                        constrain_slot!(
+                            concrete_path,
+                            hash2_count,
+                            preimg,
+                            img,
+                            store.poseidon_cache.constants.c4()
+                        );
+                        hash2_count += 1;
+                    }
+                    HashArity::A3 => {
+                        constrain_slot!(
+                            concrete_path,
+                            hash3_count,
+                            preimg,
+                            img,
+                            store.poseidon_cache.constants.c6()
+                        );
+                        hash3_count += 1;
+                    }
+                    HashArity::A4 => {
+                        constrain_slot!(
+                            concrete_path,
+                            hash4_count,
+                            preimg,
+                            img,
+                            store.poseidon_cache.constants.c8()
+                        );
+                        hash4_count += 1;
+                    }
+                };
+            }
+        }
+        macro_rules! fill_dummies {
+            (
+                $lower: expr,
+                $upper: expr,
+                $preimg_size: expr,
+                $constants: expr
+            ) => {
+                for item in $lower..$upper {
+                    constrain_slot!(
+                        &Boolean::Constant(false),
+                        item,
+                        vec![alloc_dummy_hash.clone(); $preimg_size],
+                        &alloc_dummy_hash,
+                        $constants
+                    );
+                }
+            };
+        }
+
+        ///////////////// Fill with dummies: /////////////////
+        fill_dummies!(
+            hash2_count,
```

Comment:
Note: `hash2_count` is the number of concrete `Hash2`s in the slot 'info's (I think).